arch/riscv/core/smp.c

```diff
@@ -54,47 +55,46 @@ void z_riscv_secondary_cpu_init(int cpu_num)
 }
 
 #ifdef CONFIG_SMP
-static uint32_t *get_hart_msip(int hart_id)
-{
-	return (uint32_t *)(unsigned long)(RISCV_MSIP_BASE + (hart_id * 4));
-}
 
-void arch_sched_ipi(void)
-{
-	unsigned int key;
-	uint32_t i;
-	uint8_t id;
+#define MSIP(hartid) ((volatile uint32_t *)RISCV_MSIP_BASE)[hartid]
 
-	key = arch_irq_lock();
+static atomic_val_t cpu_pending_ipi[CONFIG_MP_MAX_NUM_CPUS];
+#define IPI_SCHED	BIT(0)
 
-	id = _current_cpu->id;
+void arch_sched_ipi(void)
+{
+	unsigned int key = arch_irq_lock();
+	unsigned int id = _current_cpu->id;
 	unsigned int num_cpus = arch_num_cpus();
 
-	for (i = 0U; i < num_cpus; i++) {
+	for (unsigned int i = 0; i < num_cpus; i++) {
 		if (i != id && _kernel.cpus[i].arch.online) {
-			volatile uint32_t *r = get_hart_msip(_kernel.cpus[i].arch.hartid);
-			*r = 1U;
+			atomic_or(&cpu_pending_ipi[i], IPI_SCHED);
+			MSIP(_kernel.cpus[i].arch.hartid) = 1;
 		}
 	}
 
 	arch_irq_unlock(key);
 }
 
-static void sched_ipi_handler(const void *unused)
+static void ipi_handler(const void *unused)
 {
 	ARG_UNUSED(unused);
 
-	volatile uint32_t *r = get_hart_msip(csr_read(mhartid));
-	*r = 0U;
+	MSIP(csr_read(mhartid)) = 0;
 
-	z_sched_ipi();
+	atomic_val_t pending_ipi = atomic_get(&cpu_pending_ipi[_current_cpu->id]);
+
+	if (pending_ipi & IPI_SCHED) {
```

Comment:
Hi, @npitre, you probably should clear corresponding bits cpu_pending_ipi now, when you have introduced multiple IPI causes. I do not see any place where these bits are cleared and handling the shed_ipi in response to FP flush IPI or vice versa is probably not what you want. This kinda worked when sched_ip was the only IPI service defined, but needs greater care now.

Reply:
Indeed, good catch. In my mind I did it but apparently not.
See PR #54116.